### ISSUE TITLE
Reducing False Positive cases

### ISSUE THREAD

**Issue by Sauravpandey98** [ultralytics/yolov5#6731]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi to all, I am working on the problem of detecting boxes in warehouses.I am using yolov5 for object detection and it is working fine in most of the cases.But I am getting a particular kind of false detections that I am not able to remove.Example of of such case is shown below.Any suggestion to solve this problem would be appreciated.

![Ground Truth](https://user-images.githubusercontent.com/62154537/155002995-f7b40d4c-55e4-4200-85ed-52c277636c18.jpeg)
![Ground Truth](https://user-images.githubusercontent.com/62154537/155003343-7add7aa8-fbab-40d7-a3a3-7d4b3ab51bc2.jpeg)

![Prediction](https://user-images.githubusercontent.com/62154537/155003153-90a86227-a89d-4a71-9710-3ee8a400f9b5.jpeg)
![Prediction](https://user-images.githubusercontent.com/62154537/155003198-77e6d3e3-1ac1-42d3-8ca4-b7f8d3ff9417.jpeg)

**Comment by glenn-jocher** [comment]
@Sauravpandey98 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by Guttappa1238** [comment]
Hi @Sauravpandey98, Give details about size of the dataset and training results/graphs if possible, so that everybody will understand your issue.

For now my suggestion is try to run the detection using [yolov5m_Objects365.pt](https://github.com/ultralytics/yolov5/releases/download/v6.0/yolov5m_Objects365.pt)  by passing --classes 20 (i.e Storage box). i hope you will get better results.

`python path/to/detect.py --weights yolov5m_Objects365.pt --classes 20`

**Comment by Sauravpandey98** [comment]
yeah thanks @Guttappa1238 and @glenn-jocher for responding.
@Guttappa1238 I have tried with yolov5m_Objects365 model .But it is not performing well.

In my dataset , I have only one class i.e "box"  as you can see from the example images given in issue.Firstly I have trained a base model using a dataset of 1200 training images and manually tuned the hyper-parameters to find best model.After this I fine tuned the best model on 1000 training images.These prediction result are using the fine tuned model.

Now In my case most of false positive is occurring when the model picks wrinkled polythene wrapping or tapes as edges.So I am asking what changes should I do in my dataset or training strategy to decrease these kind of false positives.

**Comment by HAN-007** [comment]
hello, can I learn the original version of the pictures and which model you are using?
(image without bbox)

**Comment by anujdutt9** [comment]
@Sauravpandey98 Couple of suggestions that might help:
1. Try to add more images if possible. Increasing the dataset size should help.
2. Make sure that your annotations are correct and don't have any False labels.
3. Try to begin with a pre-trained model provided in this repository like a COCO pre-trained YOLOv5s/m/l model and finetune it with your dataset as a whole instead of in 2 phases.
4. Once you have a good base model then try to make changes to the hyperparameters to see which ones give you the most gains. 
5. Under `utils/augmentations.py` try to play around with augmentations, specifically under albumentations class. One example could be adding random sun flare, that could mimic the lights falling on the tapes on the boxes that might be a potential cause for the issue. https://albumentations.ai/docs/api_reference/augmentations/transforms/#albumentations.augmentations.transforms.RandomSunFlare
6. The default fitness will favor mutations that produce higher mAP@0.5:0.95 mostly. You can shift this to F1 or mAP@0.5 to reduce the influence of accurate bounding boxes by modifying the fitness function here: 
https://github.com/ultralytics/yolov5/blob/747c2653eecfb870b1ed40b1e00e0ef209b036e9/utils/metrics.py#L12-L16

Hope this helps.

**Comment by Sauravpandey98** [comment]
@anujdutt9 thanks for suggestions.But I have a query regarding mAP values.So on observation I found that even if the model starts overfitting ( generally it occurs when object loss starts increasing),the mAP@0.5:0.95 continues to increase while the mAP@0.5 remains stable.So suppose If I give same weightage to both the mAP values to calulate fitness,the best fitness would still not be corresponds to minimum loss value.
So should I prefer fitness value here or should I use those weights that has max mAP value along with minimum loss.For example in this case:
![W B Chart 10_03_2022, 11_53_39](https://user-images.githubusercontent.com/62154537/157602304-debaedde-5a47-4c96-9b3d-5dc70f899793.png)
![W B Chart 10_03_2022, 11_53_21](https://user-images.githubusercontent.com/62154537/157602330-1152d018-c9e2-4a72-9121-bf7dc4268e3c.png)
![W B Chart 10_03_2022, 11_53_00](https://user-images.githubusercontent.com/62154537/157602345-c214027e-b826-46eb-b065-133b9cce1df8.png)
![W B Chart 10_03_2022, 11_52_09](https://user-images.githubusercontent.com/62154537/157603413-3a29db07-9a24-4e12-adfc-a94d1c01c603.png)

**Comment by glenn-jocher** [comment]
@Sauravpandey98 sure. Some metrics like mAP@0.5:0.95 are more influenced by regression loss than obj loss and vice versa. The default fitness function is most weighted towards the main COCO metric of mAP@0.5:0.95:
https://github.com/ultralytics/yolov5/blob/6dd82c025298d219a1eb1fe8e486fb99d5324d34/utils/metrics.py#L15-L19

Ideally all losses would begin to overfit in concert, but in practice each loss will overfit at a different point in training (assuming you train long enough). You can try to balance these manually using the loss gain hyperparameters, or evolve hyperparameters:
https://github.com/ultralytics/yolov5/blob/6dd82c025298d219a1eb1fe8e486fb99d5324d34/data/hyps/hyp.scratch-low.yaml#L13-L17

**Comment by Sauravpandey98** [comment]
Thanks @glenn-jocher for these insights into loss values.But can you please give me a suggestion?
So from the graphs you can see that object loss value started to go up from 100 epoch.While according to above mentioned fitness score I am getting best weights at 196 epoch.

So,should I use weights at 100 epoch or weights at 196 epoch for prediction?

**Comment by glenn-jocher** [comment]
@Sauravpandey98 not sure I understand your question. If you want the weights that produced the best mAP on the validation set use best.pt. That's the entire point of best.pt.
